x/auth/tx/builder.go

```diff
@@ -109,25 +104,19 @@ func (w *wrapper) GetSigners() []sdk.AccAddress {
 }
 
 func (w *wrapper) GetPubKeys() []crypto.PubKey {
-	if w.pubKeys == nil {
-		signerInfos := w.tx.AuthInfo.SignerInfos
-		pubKeys := make([]crypto.PubKey, len(signerInfos))
+	signerInfos := w.tx.AuthInfo.SignerInfos
+	pks := make([]crypto.PubKey, len(signerInfos))
 
-		for i, si := range signerInfos {
-			var err error
-			pk := si.PublicKey
-			if pk != nil {
-				pubKeys[i], err = w.pubkeyCodec.Decode(si.PublicKey)
-				if err != nil {
-					panic(err)
-				}
-			}
+	for i, si := range signerInfos {
+		pk, ok := si.PublicKey.GetCachedValue().(crypto.PubKey)
```

Comment:
Maybe add a check to make sure `si.PublicKey` is not nil before calling `GetCachedValue()` on it?